Add a dedicated widget for transient notifications

dragunoff opened the pull request:
I started this with the simple idea of moving the UI feedback notifications out of the chat and into a dedicated "transient messages" panel. However as I wanted to reuse the existing chat overlay widget and I wanted to have flexibility with regards to styling I ended up reworking how chat lines are displayed. The commit messages should give more detail into the changes but basically text notifications are made entirely template driven. The last couple of commits are mostly polishing. Here are some screenshots of how things look:

![Screenshot_20210621_153224](https://user-images.githubusercontent.com/1355810/122763692-4c83a780-d2a7-11eb-8ba7-ede65501a680.png)
_The new transient notifications panel_

![image](https://user-images.githubusercontent.com/1355810/122763335-e4cd5c80-d2a6-11eb-831f-ee01610c41e7.png)
_Two screenshots overlaid to show the difference between open and closed in-game chat (only full messages are displayed in the chat overlay)_

![Screenshot_20210621_145931](https://user-images.githubusercontent.com/1355810/122762716-46d99200-d2a6-11eb-914a-53ec64898d6f.png)
_Various notifications in the lobby chat_

![Screenshot_20210621_153037](https://user-images.githubusercontent.com/1355810/122763608-31189c80-d2a7-11eb-92d6-be8a9335dfa6.png)
_In-game menu chat_

Depends on #18905

dragunoff committed:
Rework chat line templates and logic

- Extract chat line templates and logic so they can be reused across widgets
- Make text notification styling entirely template driven (by removing chat color configuration and making color optional for `TextNotification`)
- Add a new TextNotificationsDisplay widget (based on and replacing ChatDisplayWidget)
- Add timestamp support to text notifications

dragunoff committed:
Add a transients panel with corresponding chrome logic and a notification template

dragunoff committed:
Move audio muted/unmuted notifications to feedback pool

dragunoff committed:
Add text feedback for type selection hotkey when selection is empty

dragunoff committed:
Add audio feedback for selection hotkeys

dragunoff committed:
Add full-stop to UI feedback text notifications

pchote commented:
Now that we have the transient display widget it would be good to add text notifications for the spammier things like base under attack / silos needed / unit lost before we ship a playtest. Is this on your roadmap already for a followup PR?

dragunoff commented:
> Now that we have the transient display widget it would be good to add text notifications for the spammier things like base under attack / silos needed / unit lost before we ship a playtest. Is this on your roadmap already for a followup PR?

Yes, I'm thinking about it but I'm taking a different approach than what I did in #18929. Instead of attaching the notifications as transcriptions to the audio I want to add them to the respective traits as independent text notifications.

pchote commented:
Needs a rebase.

dragunoff commented:
Update: rebased and resolved conflicts.

dragunoff commented:
I know this overlaps the performance graph but I will think about a fix if the general approach here is approved.

pchote commented on OpenRA.Game/TextNotification.cs:
```diff
@@ -21,16 +21,18 @@ public class TextNotification : IEquatable<TextNotification>
 			Text = text;
 			PrefixColor = prefixColor;
 			TextColor = textColor;
+			Time = DateTime.Now;
```
I imagine this will break if we ever want to expose timestamps for replays or savegames, but there are bigger issues blocking that than this. Not asking you to change anything, I just thought it important to raise this.

dragunoff replied:
I didn't consider this when adding the timestamps to the in-game menu chat. This is only necessary for the lobby, so I'll remove the timestamps from the in-game chat.

pchote replied:
Its fine for the ingame menu chat too, thats not visible in replays or singleplayer (and we dont support multiplayer saves). IMO we should keep it.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs:
```diff
@@ -263,53 +267,31 @@ public void AddChatLineWrapper(TextNotification chatLine)
-			textLabel.GetText = () => text;
-			var dh = font.Measure(text).Y - textLabel.Bounds.Height;
-			if (dh > 0)
+			var chatLine = Ui.LoadWidget(templates[notification.Pool.ToString()], null, new WidgetArgs
```
`Ui.LoadWidget` uses a significant amount of reflection, so i expect it to be orders of magnitude slower than `chatTemplate.Clone()`. Can we find a way to avoid this?

dragunoff replied:
Is something like this better:
```csharp
// In the ctor
if (logicArgs.TryGetValue("Templates", out var templateIds))
	foreach (var item in templateIds.ToDictionary(kv => kv.Value))
		templates[item.Key] = Ui.LoadWidget(item.Value, null, new WidgetArgs());

// and then when adding the lines
var chatLine = templates[notification.Pool.ToString()].Clone();
new TextNotificationLogic(chatLine, notification, boxWidth, withTimestamp);
```

I tested and it works but vscode is flagging the "new instance of TextNotificationLogic which is never used".

dragunoff replied:
Or maybe handle this with a static method on the `TextNotificationLogic` class (or elsewhere)? Something like:
```csharp
// when adding the lines
var chatLine = templates[notification.Pool.ToString()].Clone();
TextNotificationLogic.ApplyTo(chatLine, notification, boxWidth, withTimestamp);
```

dragunoff replied:
Added a static method to `WidgetUtils`.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -489,13 +489,17 @@ public override void Tick()
 		{
-			var template = (ContainerWidget)chatTemplate.Clone();
-			LobbyUtils.SetupChatLine(template, DateTime.Now, chatLine);
+			var chatLine = Ui.LoadWidget(chatTemplates[notification.Pool.ToString()], null, new WidgetArgs
```
Likewise here.

pchote commented on OpenRA.Mods.Common/Widgets/TextNotificationsDisplayWidget.cs:
```diff
@@ -0,0 +1,160 @@
+
+		public void AddNotification(TextNotification notification)
+		{
+			var notificationWidget = Ui.LoadWidget(templates[notification.Pool.ToString()], null, new WidgetArgs
```
And here.

pchote commented on OpenRA.Mods.Common/Widgets/TextNotificationsDisplayWidget.cs:
```diff
@@ -0,0 +1,160 @@
+		public string SystemTemplate = "SYSTEM_LINE_TEMPLATE";
+		public string MissionTemplate = "SYSTEM_LINE_TEMPLATE";
+		public string FeedbackTemplate = "SYSTEM_LINE_TEMPLATE";
+		readonly Dictionary<string, string> templates = new Dictionary<string, string>();
```
Why not `<TextNotificationPool, string>`?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameTransientNotificationsLogic.cs:
```diff
@@ -0,0 +1,98 @@
+
+			var lineToDisplay = notification;
+
+			if (displayWidget.Children.Count > 0 && notification.CanIncrementOnDuplicate() && notification.Equals(lastLine))
```
`notification == lastLine`?

dragunoff replied:
That doesn't work. I guess because the `==` operator is not overloaded.

pchote replied:
That sounds plausible. The standard pattern is to implement the actual equality logic in the `operator ==` definition, then have the `IEquatable.Equals` implementation just check `this == other`.

pchote reviewed (commented):
General approach looks sensible on a first pass:

dragunoff commented:
Update: rebased and resolved review comments. 

Also moved the perf graph out of the way:
![image](https://user-images.githubusercontent.com/1355810/134782320-a0990337-23f1-4c4a-8efb-9134b9f827af.png)

dragunoff commented:
Update: oof, also resolved https://github.com/OpenRA/OpenRA/pull/19458#discussion_r663549087

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs:
```diff
@@ -60,11 +64,15 @@ public IngameChatLogic(Widget widget, OrderManager orderManager, World world, Mo
 
+			if (logicArgs.TryGetValue("Templates", out var templateIds))
+				foreach (var item in templateIds.ToDictionary(kv => kv.Value))
+					templates[item.Key] = Ui.LoadWidget(item.Value, null, new WidgetArgs());
```
How about something like this?
```csharp
if (logicArgs.TryGetValue("Templates", out var templateIds))
{
	foreach (var item in templateIds.Nodes)
	{
		var key = FieldLoader.GetValue<TextNotificationPool>("key", item.Key);
		templates[key] = Ui.LoadWidget(item.Value.Value, null, new WidgetArgs());
	}
}
```

This would avoid the `ToDictionary` call, and allows that `templates` to key on `TextNotificationPool` to save the `ToString`s when adding lines and better match `TextNotificationsDisplayWidget`.

(and likewise in `LobbyLogic`).

pchote commented on OpenRA.Mods.Common/Widgets/TextNotificationsDisplayWidget.cs:
```diff
@@ -0,0 +1,157 @@
+			var lastChildOverflows = Bounds.Height < Children[Children.Count - 1].Bounds.Height;
+
+			if (HideOverflow && lastChildOverflows)
+				Game.Renderer.EnableScissor(GetDrawBounds());
```
`GetDrawBounds` looks relately expensive. Would it make sense to wrap this in a `CachedTransform` or cache the new height manually when adding/removing notifications?

pchote replied:
Or, hopefully better, can we avoid the scissor completely by only calling a subset of `child.DrawOuter` below?

dragunoff replied:
> Or, hopefully better, can we avoid the scissor completely by only calling a subset of `child.DrawOuter` below?

That's a good idea - will look into it. The scissor would still be necessary when there's a single message that overflows because then I want to display at least the part of it that fits in the widget.

pchote replied:
I thought we were already adjusting the layout to only ever draw full lines?

Edit: Oh, I guess this is when messages are wrapped over multiple lines. Would it make sense to break each line into its own widget?

dragunoff replied:
This is only with regards to the extreme case where the latest message overflows our desired widget max height:
![Screenshot_20210927_161707](https://user-images.githubusercontent.com/1355810/134916927-a19e13d5-2e24-48dd-b85a-3c74e3c97a2e.png)
_Scissors in action_

![Screenshot_20210927_161755](https://user-images.githubusercontent.com/1355810/134917001-66c55fe9-2fef-4134-be40-0ef73734eff2.png)
_The full message is visible in the log_

I don't want to be displaying this (or even longer messages if anyone decides to abuse it):
![image](https://user-images.githubusercontent.com/1355810/134917849-84730175-8c36-4618-a5bd-bbec20b7d89a.png)

I am already avoiding the scissor most of the time - it is only ever used if `HideOverflow` is set to `true` and the last child overflows the desired bounds. The draw bands shouldn't have to be calculated over and over though - the number of lines we can fit in the widget is always the same so this can be calculated once. There is some left-over logic from an earlier iteration in the `GetDrawBounds` method - I will clean it up.

dragunoff replied:
OK, the scissor bounds are now calculated only once at widget initialization. This is assuming that all templates use the same font which I dare say would be the case 99% of the time.

pchote reviewed (commented):
Ingame behaviour and overall code structure is looking good. Just a couple of localized points below:

dragunoff commented:
Update: rebased and addressed https://github.com/OpenRA/OpenRA/pull/19458#discussion_r716099248 and https://github.com/OpenRA/OpenRA/pull/19458#discussion_r716099594

dragunoff commented:
Update: change `GetHashCode()` of `TextNotification` as suggested by @teinarss and add explicit height to the transients widget (this is necessary with the new overflow bounds calculations).

teinarss commented on OpenRA.Mods.Common/Widgets/TextNotificationsDisplayWidget.cs:
```diff
@@ -0,0 +1,136 @@
+				Children[i].Bounds.Y += mostRecentChild.Bounds.Height + ItemSpacing;
+		}
+
+		public void RemoveNotification()
```
Can be private

teinarss commented on OpenRA.Mods.Common/Widgets/WidgetUtils.cs:
```diff
@@ -364,6 +364,62 @@ public static void BindPlayerNameAndStatus(LabelWidget label, Player p)
+			textLabel.Bounds.Width = textWidth;
+
+			// Hack around our hacky wordwrap behavior: need to resize the widget to fit the text
+			var text = WidgetUtils.WrapText(notification.Text, textLabel.Bounds.Width, textFont);
```
No need to use `WidgetUtils.`

dragunoff commented:
Rebased.

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/6bf1f74e5fc67fd96cea2d72e0050be8dfa10871...ae8bc2a26b4b8150afdad20fb23a6a8ea27b2f9f)
